### PR TITLE
add NatGateway datasource

### LINKED ISSUE #475: Feature request: NAT data source and more friendly elastic load balance resource definition
Hiii,
Unless the Terraform documentation is missing a part, it seems that there isn't a data source for NAT gateway? This data source will be vital for our use case where we already purchased a shared NAT beforehand and would like use it to provide internet access to multiple clusters. Currently, for the lack of NAT data source, we have to hardcode the gateway ID in the SAT rule definition like: 

```
resource "huaweicloud_nat_snat_rule_v2" "snat" {
  # TODO: read it from data source!
  nat_gateway_id = "xxx-xxx-xxx-xxx"
  network_id     = huaweicloud_vpc_subnet_v1.subnet.id
  floating_ip_id = huaweicloud_vpc_eip_v1.nat_ip.id
}
```

Another petition is in regards with the [Elastic Load Balancer resource API](https://www.terraform.io/docs/providers/huaweicloud/r/lb_loadbalancer_v2.html), which looks quite confusing to us, and perhaps a bit outdated already? The example gives a "floating_ip" which it is undocumented. In a nutshell, we hope to create a ELB similar to how the portal guided us --- select a flavour (dedicated or shared), elastic IP (use an existing one or create a new one) and VPC & subnet. 

```
resource "huaweicloud_lb_loadbalancer_v2" "lb_1" {
  vip_subnet_id = "d9415786-5f1a-428b-b35f-2f1523e146d2"
}

resource "huaweicloud_networking_eip_associate" "eip_1" {
  floating_ip = "1.2.3.4"
  port_id     = "${huaweicloud_lb_loadbalancer_v2.lb_1.vip_port_id}"
}
```

Thanks heaps!

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
make testacc TEST='./huaweicloud' TESTARGS='-run TestAccNatGatewayDataSource_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccNatGatewayDataSource_basic -timeout 360m
=== RUN   TestAccNatGatewayDataSource_basic
--- PASS: TestAccNatGatewayDataSource_basic (87.17s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       87.207s

fixes #475

### CODE REVIEW

File: website/docs/d/nat_gateway.html.markdown [hunk 1]
```diff
@@ -0,0 +1,78 @@
+---
```
Comment (niuzhenguo) [inline website/docs/d/nat_gateway.html.markdown:1]: need to add `subcategory: "NAT Gateway (NAT)"`

File: website/docs/d/nat_gateway.html.markdown [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+layout: "huaweicloud"
+page_title: "HuaweiCloud: huaweicloud_nat_gateway"
+sidebar_current: "docs-huaweicloud-datasource-nat-gateway"
+description: |-
+  Get information on a HuaweiCloud NAT gateway.
+---
+
+# huaweicloud\_networking\_network\_v2
```
Comment (niuzhenguo) [inline website/docs/d/nat_gateway.html.markdown]: huaweicloud\_nat\_gateway

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+resource "huaweicloud_networking_subnet_v2" "subnet_1" {
+  cidr = "192.168.199.0/24"
+  ip_version = 4
+  network_id = "${huaweicloud_networking_network_v2.network_1.id}"
+}
+
+resource "huaweicloud_networking_router_interface_v2" "int_1" {
+  subnet_id = "${huaweicloud_networking_subnet_v2.subnet_1.id}"
+  router_id = "${huaweicloud_networking_router_v2.router_1.id}"
+}
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: these networking resources are all deprecated, use vpc and vpc_subnet resources instead.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+resource "huaweicloud_nat_gateway_v2" "nat_1" {
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: use `huaweicloud_nat_gateway` instead

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  name   = "%s"
+  description = "test for terraform"
+  spec = "1"
+  internal_network_id = "${huaweicloud_networking_network_v2.network_1.id}"
+  router_id = "${huaweicloud_networking_router_v2.router_1.id}"
+  depends_on = ["huaweicloud_networking_router_interface_v2.int_1"]
+}
+
+data "huaweicloud_nat_gateway" "nat_by_name" {
+	name = huaweicloud_nat_gateway_v2.nat_1.name
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: ditto.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+data "huaweicloud_nat_gateway" "nat_by_id" {
+	id = huaweicloud_nat_gateway_v2.nat_1.id
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: ditto.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/huaweicloud/golangsdk/openstack/networking/v2/extensions/natgateways"
+)
+
+func dataSourceNatGatewayV2() *schema.Resource {
+	return &schema.Resource{
+		Read:               dataSourceNatGatewayV2Read,
+		DeprecationMessage: "use huaweicloud_vpc_subnet_v1 data source instead",
+
+		Schema: map[string]*schema.Schema{
+			"limit": {
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2.go]: `limit` is unneeded. it's useless for us.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/huaweicloud/golangsdk/openstack/networking/v2/extensions/natgateways"
+)
+
+func dataSourceNatGatewayV2() *schema.Resource {
+	return &schema.Resource{
+		Read:               dataSourceNatGatewayV2Read,
+		DeprecationMessage: "use huaweicloud_vpc_subnet_v1 data source instead",
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2.go]: remove this DeprecationMessage.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		Schema: map[string]*schema.Schema{
+			"limit": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"tenant_id": {
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2.go]: `tenant_id` is unneeded here.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+			},
+			"admin_state_up": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"region": {
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2.go]: `region` is unneeded.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"admin_state_up": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"created_at": {
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2.go]: `createdat` is unneeded.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+		TenantId:          d.Get("tenant_id").(string),
+		Description:       d.Get("description").(string),
+		Spec:              d.Get("spec").(string),
+		RouterID:          d.Get("router_id").(string),
+		InternalNetworkID: d.Get("internal_network_id").(string),
+		Status:            d.Get("status").(string),
+		CreatedAt:         d.Get("created_at").(string),
+	}
+
+	if v, ok := d.GetOk("status"); ok {
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2.go]: why not just passing the status argument as above.

**Review by niuzhenguo — CHANGES_REQUESTED** [review]

File: website/docs/d/nat_gateway.html.markdown [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+* `status` - (Optional) The status of the NAT gateway.
+
+* `created_at` - (Optional) When the NAT gateway is created (UTC time). Its value rounds to 6 decimal places for seconds. The format is yyyy-mm-dd hh:mm:ss.
+
+## Attributes Reference
+
+The following attributes
+are exported:
+
+* `id` - (Optional) The ID of the NAT gateway.
```
Comment (ShiChangkuo) [inline website/docs/d/nat_gateway.html.markdown]: No need to add `optional` or `required` tags on attributes.

File: website/docs/d/nat_gateway.html.markdown [hunk 14]
```diff
@@ -0,0 +1,72 @@
+subcategory: "NAT Gateway (NAT)"
```
Comment (niuzhenguo) [inline website/docs/d/nat_gateway.html.markdown]: this should be under "---"

File: website/docs/d/nat_gateway.html.markdown [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+* `status` - (Optional) The status of the NAT gateway.
+
+
+## Attributes Reference
+
+The following attributes
+are exported:
+
+* `id` - The ID of the NAT gateway.
```
Comment (niuzhenguo) [inline website/docs/d/nat_gateway.html.markdown]: Attributes can be only ones that not included in `Argument Reference`

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+	gateway_ip = "192.168.199.1"
+	vpc_id     = huaweicloud_vpc.vpc_1.id
+	}
+
+resource "huaweicloud_nat_gateway" "nat_1" {
+	name   = "%s"
+	description = "test for terraform"
+	spec = "1"
+	internal_network_id = "${huaweicloud_vpc_subnet.subnet_1.id}"
+	router_id = "${huaweicloud_vpc.vpc_1.id}"
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: just use `router_id = huaweicloud_vpc.vpc_1.id`

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+		}
+
+		return nil
+	}
+}
+
+func testAccNatGatewayV2DataSource_basic(name string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_vpc" "vpc_1" {
+	name = "vpc_1"
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: all name should use the value of the `name` argument.

File: huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+	cidr = "192.168.0.0/16"
+	}
+
+resource "huaweicloud_vpc_subnet" "subnet_1" {
+	name       = "subnet_1"
```
Comment (niuzhenguo) [inline huaweicloud/data_source_huaweicloud_nat_gateway_v2_test.go]: ditto.

**Review by niuzhenguo — CHANGES_REQUESTED** [review]

**Review by niuzhenguo — APPROVED** [review]

**Pull request merged**